### PR TITLE
Form icons

### PR DESCRIPTION
**Pull request by benrudolph**
@dimagi/product @orangejenny @biyeun  Icons for forms in the app builder

screen shot 2016-07-22 at 4 24 37 pm
screen shot 2016-07-22 at 4 24 29 pm

This was hugely helpful in me understanding my own app structure. Granted I'm not the target user, but strongly think this is a win. Realize we may argue on icons. Some other possibilities for form open/update:

Open:  
screen shot 2016-07-22 at 4 28 25 pm
screen shot 2016-07-22 at 4 29 41 pm

Update:
screen shot 2016-07-22 at 4 29 08 pm

### CODE REVIEW

File: corehq/apps/app_manager/models.py [hunk 1]
```diff
@@ -1316,6 +1316,28 @@ def _get_active_actions(self, types):
...
+    @memoized
+    def get_icon_help_text(self):
+        subcases = self.get_subcase_types()
+        messages = []
+
+        if subcases:
+            messages.append(_('This form opens cases for {}').format(', '.join(subcases)))
+
+        if self.requires_case():
+            messages.append(('This form updates the {} case').format(self.get_module().case_type))
```
Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: translation?

Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: I think it's not unusual to use "case" for the case type, which makes this `This form updates the case case.` How about `This form will update a {}`?

File: corehq/apps/app_manager/models.py [hunk 2]
```diff
@@ -1316,6 +1316,28 @@ def _get_active_actions(self, types):
...
+        else:
+            return 'none'
+
+    @memoized
+    def get_icon_help_text(self):
+        subcases = self.get_subcase_types()
+        messages = []
+
+        if subcases:
+            messages.append(_('This form opens cases for {}').format(', '.join(subcases)))
```
Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: `This form opens a {}`?

File: corehq/apps/app_manager/models.py (hunk 1 above)
Comment (benrudolph) [inline corehq/apps/app_manager/models.py]: mmm

Comment (benrudolph) [inline corehq/apps/app_manager/models.py]: good point. also missed the underscore thanks

**Comment by orangejenny** [comment]
Hey, I think that
a) This is great
b) We should be very deliberate about both the icons & the text

File: corehq/apps/app_manager/models.py (hunk 1 above)
Comment (benrudolph) [inline corehq/apps/app_manager/models.py]: maybe keep it in present tense with `This form updates a {}`? so it matches the open?

**Comment by orangejenny** [comment]
Any reason for leaving off case closing? Also might want to discuss if we should differentiate between opening a case and opening a child case.

**Comment by benrudolph** [comment]
just to keep it very simple at the beginning, though i could see closing a case being useful too

**Comment by benrudolph** [comment]
added close case 
screen shot 2016-07-22 at 5 03 36 pm

**Comment by orangejenny** [comment]
The app summary uses envelopes to represent cases, though I don't know how deliberate of a decision that was. I'm not crazy about either of the icons in the first screenshot.

Open: both of your additional suggestions seem like possibilities. Or maybe screen shot 2016-07-22 at 5 30 34 pm
Update: A refresh-type icon? Maybe screen shot 2016-07-22 at 5 31 39 pm
Close: eject? Not great. But I'm starting to think that the thing we should get across is that the data moves off of the phone. What about doing three custom icons, each with a phone: "open" with an arrow going into the phone, "update" with an arrow circling back to the phone, and "close" with an arrow exiting the phone.

Another idea that definitely isn't right, but I'm thinking we might want three icons that are obviously related in some way: hourglasses?
screen shot 2016-07-22 at 5 32 36 pm
screen shot 2016-07-22 at 5 32 40 pm
screen shot 2016-07-22 at 5 32 48 pm

File: corehq/apps/app_manager/models.py (hunk 1 above)
Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: Yeah, consistent tense is good.

File: corehq/apps/app_manager/models.py [hunk 3]
```diff
@@ -1316,6 +1316,28 @@ def _get_active_actions(self, types):
                 actions[action_type] = a
         return actions
 
+    @memoized
+    def get_action_type(self):
+        if self.get_subcase_types():
```
Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: It seems weird to me that a method called `get_subcase_types` gets cases opened by the form (as opposed to child cases opened by the form) but it looks like that is indeed the way it is.

**Comment by benrudolph** [comment]
dig the `sign-in` as the create.
i'm wishy washy on `refresh` as i think we use that in a lot of places on the site

also think there's a lot of potential to the hourglass idea

File: corehq/apps/app_manager/models.py (hunk 3 above)
Comment (benrudolph) [inline corehq/apps/app_manager/models.py]: (updated to be more readable)

**Comment by benrudolph** [comment]
did some thinking about your hourglass idea. i'm starting to move a bit away from it now. i feel like something like that is similar to a "sequential" visualization (e.g some sort of hue of the same color getting dark to represent more of one thing). and while the sequential view is does bring out the life cycle of the case a bit more i don't necessarily think that's how all users think about it. especially in a non linear application. the sequential idea also loses some of the differentiating ability; open, update, close will be harder to discern from one another. 

i'm now more of a fan of imposing "categorical" icons where the icons are representative of the actions, but sufficiently distinct so that one can clearly tell the difference between two at a glance

**Comment by amsagoff** [comment]
Hey,

This is a cool idea - let's take the spec-ing of what icons to use
somewhere else? Ben, is it straightforward to throw a spec doc together
that we can iterate on? We have a ton of icons already on this page and
throughout HQ that mean different things, so I think getting the icons
right/not overloading a single icon with too much meaning will be important.

**Comment by benrudolph** [comment]
Sounds good, will do

**Comment by biyeun** [comment]
I'm in agreement with taking a closer look and possibly creating some custom icons. Will add this to review tomorrow

**Comment by orangejenny** [comment]
Hey, I'm wondering if this should go behind the support flag. The major use case I can think of is orienting yourself to an unfamiliar app, which is something support/devs do all the time, but real users don't do often. The icons are potentially a learning aid for new users, but you could also argue that we want users to think *less* about cases and shouldn't be more obviously surfacing case actions.

I don't want to get deeply into this topic here, since we're meeting to talk about icons anyway, but wanted to bring it up before we meet.

**Comment by benrudolph** [comment]
external-link
undo
ban

**Comment by benrudolph** [comment]
Ok threw this behind the support toggle. I'm not gonna push on this for using it on all users, but I love this feature and maybe the correct home for it is under the support toggle. Let me know if there's any other pending feedback, else this is g2g

File: corehq/apps/app_manager/templates/app_manager/managed_app.html [hunk 4]
```diff
@@ -207,6 +208,35 @@
                                         <!--[F]-->
                                         <a id="view_form_{{ module.id }}_{{ form.id }}_sidebar" href="{% url "view_form" domain app.id module.id form.id %}">
                                             <i class="drag_handle"></i>
+                                            {% if request|toggle_enabled:"SUPPORT" %}
+                                            {% if form.get_action_type == 'open' %}
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/managed_app.html]: minor, but this block could be DRYer

**Comment by amsagoff** [comment]
Hey,

I think you should talk to support before concluding it belongs behind the
support toggle. There are also downsides to having your app look
significantly different for you vs the rest of the world, especially if
you're on the support toggle but are also an FM.

Ideally, this would live behind a toggle only temporarily til we figure out
the right long-term place for it. I think if you want to deploy this to HQ
you'll need to drive it through the rest of the process as well. (Trying to
have our feature flag list be less of a graveyard).

File: corehq/apps/app_manager/templates/app_manager/managed_app.html (hunk 4 above)
Comment (benrudolph) [inline corehq/apps/app_manager/templates/app_manager/managed_app.html]: yeah good point

**Comment by emord** [comment]
Maybe a crazy idea: what if we just passed in more info about the app (or maybe the whole app doc) for support users. Then we could push creating the icons to a greasemonkey script until there's a consensus on that? That way it's easy to turn on/off without turning off the support flag and could make other changes like this easier to prototype on prod

**Comment by benrudolph** [comment]
Talked this over with Robin and Amy:

Robin - says she's onboard, preferred to drop the icon for the form with no case mgmt. i ended up doing that suggestion
Amy - she thought it was very useful and something that a lot of GS people would like a lot. and, surprisingly she was able to name what each icon did without prompting. she also had a bunch of other ideas, but i feel like they're out of scope

✅ 

@emord i like that idea though i think it's much more work for this and would be a bigger lift for people to use it if they weren't familiar with greasemonkey

**Comment by benrudolph** [comment]
> Ideally, this would live behind a toggle only temporarily til we figure out
the right long-term place for it. I think if you want to deploy this to HQ
you'll need to drive it through the rest of the process as well. (Trying to
have our feature flag list be less of a graveyard).

I think for now support toggle is fine. I want it to be useable for at least some people and I don't think I have the bandwidth to try and get consensus on which icons to use if we'd release it to the public. If there's a good reaction to it, then maybe I can revisit

**Comment by benrudolph** [comment]
@amsagoff are you signed off with this?

**Comment by amsagoff** [comment]
I think let's go to voice on this quickly:

"I think for now support toggle is fine. I want it to be useable for at
least some people and I don't think I have the bandwidth to try and get
consensus on which icons to use if we'd release it to the public. If
there's a good reaction to it, then maybe I can revisit"

**Comment by benrudolph** [comment]
offlined with amelia and we're good to merge this with the stipulation that this is for support and doesn't have the intention of being driven to become part of the product. it has been verified that this is useful for support, so I think fine to leave

**Comment by czue** [comment]
> doesn't have the intention of being driven to become part of the product

Is there a quick summary of why this was the decision? Mostly curious if you all think the feature is not valuable or it's more of a bandwidth thing.

**Comment by benrudolph** [comment]
Here's what I understand ( @amsagoff feel free to fill in any bits i missed/misunderstood ): essentially product is making an effort to cut down on features and saying no to more stuff that hasn't been proven to be useful. i don't think we have a great way of proving a feature useful for a small thing like this and i think that would take leg work on my part to find a way to prove that this is useful. right now i don't have the time to figure that out and amelia also has many features on her plate and also doesn't have the time to push this through to the end. as a side note, we did have a meeting on the icons themselves, but it was inconclusive so i think it'd have to go beyond just having a meeting

**Comment by benrudolph** [comment]
there's also been a push to eliminate features that just live in feature flag world which accumulates tech debt which is why there was hesitancy to let this live under the support toggle

**Comment by czue** [comment]
thanks makes sense on the reasoning.

> there's also been a push to eliminate features that just live in feature flag world which accumulates tech debt which is why there was hesitancy to let this live under the support toggle

no need to respond, but i'd be interested to hear more about this if this is consensus coming out of the team. it's the first I've heard it mentioned since the aotw we did a while back and I think should be a team conversation if that's a direction we think we need to go (I'm not sure how I feel about it and don't think it's as clear-cut a situation as you mentioned in my head)

**Comment by amsagoff** [comment]
Hey Cory,

I was sharing the perspective came more from our (still open) thread about
the feature flag graveyard, and making sure we have a long-term plan for
anything we put there. Ben had conversations with GS and with support that
validated this is a useful feature from that perspective, so the support
toggle could be a comfortable long-term home if we don't end up picking
this back up again.

I agree that we haven't come to a conclusion on that thread or in team
discussions of how to approach feature flag tech debt, aside from the sense
that it is a current problem.

Amelia

**Comment by czue** [comment]
Thanks. Keen to continue the flag discussion but not here.

I had two random "lying in bed" thoughts about this feature over the weekend:

- burying it somewhere in user-defined settings might be a good temporary home for it if we did want a wider-than-toggle release (it's not a feature preview, but it would be nice to expose to non-dimagis)
- incorporating it into the app summary for everyone seems like a win to me.

**Pull request merged**